### PR TITLE
TLS codec cleanup

### PR DESCRIPTION
**Pull request by franziskuskiefer**
Sorry for the large PR but this had to happen in one go.
Please see the [tls_codec] crate for details on the tls codec implementation.

This PR removes the tls-codec and tls-codec-derive crates from the repository (they live in [tls_codec] now) as well as the codec module from the library.

Many structs and enum implement serialization through the derive macros `TlsSize`, `TlsSerialize`, and `TlsDeserialize`.
Others require manual implementation of the corresponding traits `Size`, `Serialize`, and `Deserialize`, providing the functions `tls_serialized_len`, `tls_serialize` and `tls_serialize_detached`, as well as `tls_deserialize`.
The implementation of [tls_codec] is more flexible than the old codec, taking `Read` and `Write` for (de)serializing, more efficient by directly encoding byte vectors, and allows us to delete a lot of code.

The treekem test vector is updated in this PR as well because it had some minor issues.

[tls_codec]: https://github.com/openmls/tls-codec/

### CODE REVIEW

**Review by kkohbrok — CHANGES_REQUESTED** [review]
Thanks for taking on this huge task! A few nits, but otherwise great work. Maybe we should at some point revisit the "config" approach for setting version and ciphersuite when deserialising `Secret`s. I think I'd sleep better if did the configuration on deserialization where possible.

One thing I noticed, that there were a lot of `FIXME: remove unwrap` throughout the code. Is that meant to happen still within this PR?

File: openmls/src/lib.rs [hunk 1]
```diff
@@ -21,13 +21,17 @@
 #![cfg_attr(docsrs, feature(doc_cfg))]
 #![forbid(unsafe_code)]
 
+// TODO: add group_context.serialized again
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply openmls/src/lib.rs:24]: Ah, I forgot to remove them... That's just a todo list that's done 😳

File: openmls/src/key_packages/mod.rs [hunk 2]
```diff
@@ -171,32 +175,33 @@ impl KeyPackage {
...
+        <Self as Verifiable>::verify_no_out(self, &self.payload.credential).map_err(|_| {
             log::error!("Key package signature is invalid.");
             KeyPackageError::InvalidSignature
         })
     }
 
     /// Compute the hash of the encoding of this key package.
     pub fn hash(&self) -> Vec<u8> {
-        let bytes = self.encode_detached().unwrap();
+        // FIXME: remove unwrap
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply openmls/src/key_packages/mod.rs:186]: #438

File: openmls/src/credentials/errors.rs [hunk 3]
```diff
@@ -1,6 +1,6 @@
 use crate::ciphersuite::*;
-use crate::codec::CodecError;
 use crate::config::ConfigError;
+use tls_codec::Error as TlsError;
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply openmls/src/credentials/errors.rs:3]: I have no opinions on the alias here. I made it `TlsCodecError`.

File: openmls/src/ciphersuite/codec.rs [hunk 4]
```diff
@@ -1,127 +1,55 @@
...
-    }
-    fn decode(cursor: &mut Cursor) -> Result<Self, CodecError> {
-        if let Ok(credential_type) = Self::try_from(u16::decode(cursor)?) {
-            Ok(credential_type)
-        } else {
-            Err(CodecError::DecodingError)
-        }
+impl tls_codec::Size for Ciphersuite {
+    fn tls_serialized_len(&self) -> usize {
+        CiphersuiteName::MLS10_128_DHKEMP256_AES128GCM_SHA256_P256.tls_serialized_len()
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply openmls/src/ciphersuite/codec.rs:12]: The function is not static, so it must be called on an object. The other option would be to manually write the value which is not much better 🤷🏻

**Review by franziskuskiefer — COMMENTED** [review]
> One thing I noticed, that there were a lot of FIXME: remove unwrap throughout the code. Is that meant to happen still within this PR?

No, they are just markers. I filed #438 to tackle this.

**Review by kkohbrok — DISMISSED** [review]
Thanks for addressing my comments! 🚢 it!

**Review by kkohbrok — APPROVED** [review]

**Pull request merged**